### PR TITLE
Update dos script

### PR DESCRIPTION
**Pull request by rebeccafair**
This is mainly to update the `euphonic-dos` script to read force constants and have more parity with the other scripts. While doing this I've ended up making some other changes:

* Small bug found in `euphonic-dispersion` and `euphonic-intensity-map`, `frequencies_unit` was being set before force constants had been used to calculate the phonon modes, meaning that the units weren't correctly changed when using force constants.
* More consistent arguments between scripts - definition of CLI parser largely moved to `utils` to avoid having to define common arguments multiple times
* Refactored script tests and regeneration of test data, so they're all in the same file and easier to manage. The regeneration of test data can now be done via pytest as it's defined as a test (that is skipped by default)
* Testing of both cases of phonon data and force constants data for each script, and between all the scripts the reading of all possible file types is tested
* Better variety/combinations of CLI args in testing
* Testing of writing the plot directly to file
* `--eb` and `--qb` added as short versions of `--energy-broadening` and `--q-broadening` as I found them quite cumbersome

### CODE REVIEW

**Comment by rebeccafair** [comment]
Coverage from running just `script_tests`, it seems there's quite a good coverage of `plot.py` already, although we aren't doing any checking of titles, ticks or axis labels...

```
Name                                                                                          Stmts   Miss Branch BrPart  Cover   Missing
-----------------------------------------------------------------------------------------------------------------------------------------
Euphonic\euphonic\__init__.py                 18      1      2      1    90%   11->14, 14
Euphonic\euphonic\_version.py                277    149    144     23    37%   75->93, 86->88, 88-95, 97->99, 99->100, 100-103, 114-129, 139-158, 164->165, 165, 167->175, 176->180, 177->178, 178, 180-211, 225->228, 230->231, 231-233, 242->243, 243, 246->247, 247, 262->263, 263, 267->295, 270->272, 272-274, 278->279, 279-284, 295-298, 310->311, 311, 324->333, 326->337, 329->330, 330-336, 346-353, 366-380, 391-402, 413-422, 434-442, 447->448, 448, 454->455, 455, 457->459, 459-470, 500-501, 509-518
Euphonic\euphonic\cli\__init__.py              0      0      0      0   100%
Euphonic\euphonic\cli\dispersion.py           35      0     10      1    98%   20->25
Euphonic\euphonic\cli\dos.py                  31      0      8      1    97%   19->22
Euphonic\euphonic\cli\intensity_map.py        59      0     16      3    96%   31->37, 46->49, 63->66
Euphonic\euphonic\cli\optimise_eta.py         50     19     12      4    63%   16-20, 58->59, 59, 65->66, 66, 74->75, 75, 78->79, 79-82, 88-121
Euphonic\euphonic\cli\show_sampling.py        51      1     22      1    97%   90->95, 95
Euphonic\euphonic\cli\utils.py               154     11     54      5    92%   31->35, 35, 39->40, 40, 49->52, 52-56, 85, 157->158, 158, 188-189, 204-205, 213->214, 214
Euphonic\euphonic\crystal.py                  53      7      0      0    87%   69, 100-101, 136-138, 150, 191
Euphonic\euphonic\data\__init__.py             0      0      0      0   100%
Euphonic\euphonic\debye_waller.py             32     19      0      0    41%   39-51, 55, 61, 64-66, 77-78, 90, 112-115, 132
Euphonic\euphonic\force_constants.py         475     45    134     14    90%   104, 117-121, 256->258, 258-270, 287-291, 296->300, 300-305, 321->327, 353->355, 373->374, 374, 376->378, 378, 385->386, 386-394, 398-414, 420->421, 421, 600->637, 646->671, 786->802, 793->794, 794, 826->827, 827, 917-922, 961-965, 1008->1009, 1009, 1174-1178, 1190
Euphonic\euphonic\io.py                       65     31     42      1    48%   15-22, 33->34, 34, 46-67, 93-96
Euphonic\euphonic\plot.py                    116     12     46      6    89%   12-17, 45-47, 65-66, 160->161, 161, 163->164, 164, 179->180, 180, 224->227, 274->275, 275, 329->330, 330
Euphonic\euphonic\powder.py                   53     53     20      0     0%   3-277
Euphonic\euphonic\qpoint_phonon_modes.py     139     40     26      3    69%   116->117, 117-121, 233->237, 237-238, 262->263, 263-271, 336-378, 443-445, 457
Euphonic\euphonic\readers\__init__.py          0      0      0      0   100%
Euphonic\euphonic\readers\castep.py          187     10     98      5    93%   57->58, 58, 59->60, 60, 77->83, 83-86, 238-239, 365->366, 366, 390->391, 391
Euphonic\euphonic\readers\phonopy.py         341     82     88     17    74%   15, 22, 39-40, 68-71, 88-89, 93->98, 106-107, 134-135, 138->139, 139-152, 178-179, 220->223, 225->228, 228, 231->232, 232, 255->256, 256, 281->282, 282, 341-359, 366-367, 374->375, 375, 414->417, 417, 476->478, 478, 500-528, 556-559, 626->630, 630->631, 631, 762->766, 764->765, 765, 771->772, 772-773, 778->784, 784, 791->793, 793-799
Euphonic\euphonic\sampling.py                 84      8     40      8    87%   45->48, 48, 88->91, 91, 149->152, 152, 184->187, 187, 234->237, 237, 287->288, 288, 317->320, 320, 355->358, 358
Euphonic\euphonic\spectra.py                 243     40     54     10    81%   77, 83, 95, 107-108, 114, 142->143, 143, 176->177, 177, 180->181, 181, 225-227, 241, 266-268, 354, 358, 362, 365->369, 369-377, 381->382, 382, 385->386, 386, 397->398, 398-403, 417, 442-444, 577, 605-607, 629->630, 630-631, 641->644, 644, 660->664, 664
Euphonic\euphonic\structure_factor.py         92     27     14      5    66%   80->81, 81-82, 89, 93, 98->100, 100, 141-146, 243->251, 246-247, 291->292, 292-300, 301->306, 306-313, 324-326, 338, 366-370, 388
Euphonic\euphonic\util.py                    147     24     54     10    78%   137->138, 138, 195->200, 200-205, 210->211, 211, 215->216, 216, 222->223, 223, 228-229, 317->318, 318, 326->329, 329, 344-354, 448->463, 461->448, 463->464, 464
Euphonic\euphonic\validate.py                 54     26     32      4    51%   42->43, 42->exit, 43, 48->49, 48->exit, 49, 82-83, 118-129, 149-153, 170-176
-----------------------------------------------------------------------------------------------------------------------------------------
TOTAL                                       2756    605    916    122    75%
```

File: euphonic/cli/dispersion.py
Comment (ajjackson) [inline euphonic/cli/dispersion.py:38]: I went with `is not None` so that `[]` would be considered "an empty list of labels" and still replace the default values (which I think are derived using symmetry?).

But this is academic really: the only place x_tick_labels can come from is Seekpath and it will always find at least Gamma and X points. Even if we allowed user-defined points, the range edges should be labelled _somehow_.

As we do not need to deal with `''`, `[]` etc. it doesn't matter whether we check truth or `not None` so this is fine (and easier to read).

File: tests_and_analysis/test/script_tests/test_dispersion.py
Comment (ajjackson) [inline tests_and_analysis/test/script_tests/test_dispersion.py:80]:
```python
else:
    atol = sys.float_info.epsilon
```
(Cleaner than setting and maybe replacing)

File: tests_and_analysis/test/script_tests/test_dispersion.py
Comment (ajjackson) [inline tests_and_analysis/test/script_tests/test_dispersion.py:104]: How confident are we that this cannot be accidentally run with some kind of well-meaning `--seriously-run-all-tests` option to pytest?

**Review by ajjackson — COMMENTED** [review]
This looks great, and a natural way to tidy up this code while sharing lots of functionality.

I made a few comments / queries. Would like to try it hands-on before hitting "approve" if that's ok; should be able to do that early next week?

**Comment by ajjackson** [comment]
Have had a play and the uniform interface across CLI tools seems to be working nicely.

I have found one warning that gets raised when using euphonic-dos but not the other tools: when setting energy units in wavenumbers with `--energy-unit '1/cm'`, I get

```
/home/adam/.conda/envs/science/lib/python3.8/site-packages/pint/quantity.py:1288: RuntimeWarning: divide by zero encountered in true_divide
  return self.__class__(other_magnitude / self._magnitude, 1 / self._units)
```

Something to do with normalisation, maybe?

File: euphonic/cli/intensity_map.py
Comment (rebeccafair) [reply euphonic/cli/intensity_map.py:161]: Indeed! Not having an `else` at the end makes me a little uneasy, but now there are tests for incorrect inputs, we should know if the behaviour changes

File: tests_and_analysis/test/script_tests/test_dispersion.py
Comment (rebeccafair) [reply tests_and_analysis/test/script_tests/test_dispersion.py:104]: I had a look, and it seems slightly convoluted to ignore the skip markers (at least in all the Pytest versions released so far) https://stackoverflow.com/questions/56078593/how-to-disable-skipping-a-test-in-pytest-without-modifying-the-code so its probably safe to say it can't be done by accident. If Pytest release a new argument that allows this we may have to rethink it...

**Comment by rebeccafair** [comment]
> I have found one warning that gets raised when using euphonic-dos but not the other tools: when setting energy units in wavenumbers with `--energy-unit '1/cm'`, I get
> 
> ```
> /home/adam/.conda/envs/science/lib/python3.8/site-packages/pint/quantity.py:1288: RuntimeWarning: divide by zero encountered in true_divide
>   return self.__class__(other_magnitude / self._magnitude, 1 / self._units)
> ```
> 
> Something to do with normalisation, maybe?

Good spot! It seems to occur on ` dos_bins = dos_bins.to('hartree').magnitude` in `calculate_dos` when converting from 0 1/cm to 0 Hartree. As this'll be a common occurence on this line, I could just catch the warning?

File: tests_and_analysis/test/script_tests/test_dispersion.py
Comment (ajjackson) [reply tests_and_analysis/test/script_tests/test_dispersion.py:105]: Ok! If it's not a built-in feature then I'm not too worried.

**Comment by ajjackson** [comment]
> Good spot! It seems to occur on ` dos_bins = dos_bins.to('hartree').magnitude` in `calculate_dos` when converting from 0 1/cm to 0 Hartree. As this'll be a common occurence on this line, I could just suppress this warning?

It's going to be awkward to check that this is always coming from a single value, or replace the `to()` syntax with something that is more careful. So this might be the practical way forward.

What bothers me is that I don't understand why any division by zero is necessary in the first place. What are these two magnitudes and why is our data on the bottom of the fraction?

**Comment by rebeccafair** [comment]
> What bothers me is that I don't understand why any division by zero is necessary in the first place. What are these two magnitudes and why is our data on the bottom of the fraction?

I think this must be an oversight by `pint`, I had a quick look into the source, but the logic doesn't seem that simple so probably isn't worth the time to look into it further... I don't think there's much we can do except raise an issue about it. But maybe it does make sense? The only way to have a 0 wavenumber would be to have an infinite wavelength...

Interestingly the following raises a ZeroDivisionError:
```
x = 0*ureg('1/cm')
x.to('hartree')
```
But `x = 0*ureg('1/cm').to('hartree')` doesn't.

**Comment by ajjackson** [comment]
I guess that's what we should do now then:
- suppress the warning
- ask Pint what's going on

worrying that it can sometimes lead to an error!

**Comment by rebeccafair** [comment]
Reported https://github.com/hgrecco/pint/issues/1214

**Review by ajjackson — APPROVED** [review]
That should do it :+1:

**Pull request merged**
